### PR TITLE
Skeleton of cockroach SQL driver.

### PR DESCRIPTION
**Pull request by petermattis**
The driver currently parses SQL on the client side, though this will
quickly be moved to the server.

### CODE REVIEW

File: sql/driver/conn.go [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+	case *parser.CreateDatabase:
+		return c.createDatabase(p, args)
+	case *parser.CreateTable:
+		return c.createTable(p, args)
+	case *parser.Delete:
+		return c.del(p, args)
+	case *parser.Insert:
+		return c.insert(p, args)
+	case *parser.Select:
+		return c.sel(p, args)
```
Comment (vivekmenezes) [inline sql/driver/conn.go]: s/del/delete
s/sel/select

**Comment by vivekmenezes** [comment]
LGTM

File: sql/driver/rows.go [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+	"database/sql/driver"
+	"io"
+)
+
+type rows struct {
+	columns []string
+	// TODO(pmattis): This is unfinished. Each row needs to have len(columns)
+	// entries.
+	rows []string
+	pos  int
```
Comment (tschottdorf) [inline sql/driver/rows.go]: what's `pos`?

File: sql/driver/conn.go [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+		pos:     -1,
+	}, nil
+}
+
+func (c *conn) update(p *parser.Update, args []driver.Value) (*rows, error) {
+	return nil, fmt.Errorf("TODO(pmattis): unimplemented: %T %s", p, p)
+}
+
+func (c *conn) useDatabase(p *parser.Use, args []driver.Value) (*rows, error) {
+	c.database = p.Name
```
Comment (tschottdorf) [inline sql/driver/conn.go]: will a `conn` not be used concurrently?

File: sql/driver/conn.go [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+	return &rows{}, nil
+}
+
+func (c *conn) lookupDatabase(name string) (uint32, error) {
+	nameKey := keys.MakeNameMetadataKey(structured.RootNamespaceID, name)
```
Comment (tschottdorf) [inline sql/driver/conn.go]: this should eventually cache. Too early for a TODO?

File: sql/driver/driver_test.go [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+		for rows.Next() {
+			var n string
+			if err := rows.Scan(&n); err != nil {
+				t.Fatal(err)
+			}
+			databases = append(databases, n)
+		}
+		_ = rows.Close()
+
+		expectedDatabases := databases[:i+1]
```
Comment (tschottdorf) [inline sql/driver/driver_test.go]: s/databases/names/

File: sql/driver/driver_test.go [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+		if !reflect.DeepEqual(expectedDatabases, databases) {
+			t.Errorf("expected %+v, but got %+v", expectedDatabases, databases)
```
Comment (tschottdorf) [inline sql/driver/driver_test.go]: `Fatalf` seems more appropriate, if you fail once then likely you're busted anyways and everything after is just confusing.

File: sql/driver/driver_test.go [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+		t.Fatal(err)
+	}
+
+	if _, err := db.Exec("USE t"); err != nil {
+		t.Fatal(err)
+	}
+	if _, err := db.Exec("CREATE TABLE users " + cols); !isError(err, "table .* already exists") {
+		t.Fatal(err)
+	}
+	if _, err := db.Exec("CREATE TABLE IF NOT EXISTS users " + cols); err != nil {
```
Comment (tschottdorf) [inline sql/driver/driver_test.go]: can the SQL parser also serve to auto-generate templates for all possible types of statements for our tests?

File: sql/driver/schema_test.go [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+			structured.ColumnType{Kind: structured.ColumnType_ENUM, Vals: []string{"a"}},
+			true,
+		},
+		{
+			"SET(b)",
+			structured.ColumnType{Kind: structured.ColumnType_SET, Vals: []string{"b"}},
+			true,
+		},
+		{
+			"INT NOT NULL",
```
Comment (tschottdorf) [inline sql/driver/schema_test.go]: no distinction between `INT NOT NULL` and `INT NULL`?

Comment (tschottdorf) [inline sql/driver/schema_test.go]: scratch that, just saw the bool.

File: sql/parser/sql.go [hunk 9]
```diff
@@ -333,313 +333,319 @@ var yyExca = []int{
 	-2, 0,
```
Comment (tschottdorf) [inline sql/parser/sql.go]: any chance of being able to introduce a generic comment line at the top of this file so that Github omits the diff?

**Comment by tschottdorf** [comment]
LGTM

File: sql/driver/result.go [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+// Author: Peter Mattis (peter@cockroachlabs.com)
+
+package driver
+
+type result struct {
+	lastInsertID int64
+	rowsAffected int64
+}
+
+func (r *result) LastInsertId() (int64, error) {
```
Comment (tamird) [inline sql/driver/result.go]: since golint is complaining, why not call this `LastInsertID`?

File: sql/driver/driver_test.go [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+	}
+	return s, db
+}
+
+func cleanup(s *server.TestServer, db *sql.DB) {
+	_ = db.Close()
+	s.Stop()
+}
+
+func isError(err error, re string) bool {
```
Comment (tamird) [inline sql/driver/driver_test.go]: this is done somewhere else, too, right? can we move it to util/testing?

File: sql/driver/conn.go (hunk 1 above)
Comment (petermattis) [inline sql/driver/conn.go]: `delete` and `select` are keywords in Go. I'm open to other suggestions. `del` and `sel` are quite ugly.

File: sql/driver/conn.go (hunk 3 above)
Comment (petermattis) [inline sql/driver/conn.go]: `database/sql/driver.Conn` is not used concurrently. Not sure if this is documented somewhere, but it is apparent from the usage in `database/sql` and from my inspection of the mysql driver implementation.

File: sql/driver/conn.go (hunk 4 above)
Comment (petermattis) [inline sql/driver/conn.go]: Yeah, too early for a TODO. Not sure if the caching would go here or elsewhere or even how it will work. Not even sure if we need caching of the `database name -> database id` mapping as usage of that mapping should be rare. We will need caching of `TableDescriptors`. @tamird is thinking hard about that.

File: sql/driver/driver_test.go (hunk 11 above)
Comment (petermattis) [inline sql/driver/driver_test.go]: It is done in `client/table_test.go` which will be going away shortly. Strange that `util/testing.go` is compiled into the `util` package. That doesn't seem right.

File: sql/driver/driver_test.go (hunk 5 above)
Comment (petermattis) [inline sql/driver/driver_test.go]: Oh. Heh. Oops.

File: sql/driver/driver_test.go (hunk 6 above)
Comment (petermattis) [inline sql/driver/driver_test.go]: Good point. Done.

File: sql/driver/driver_test.go (hunk 7 above)
Comment (petermattis) [inline sql/driver/driver_test.go]: Hmmm. I'm not sure. When I try to imagine this it doesn't seem any simpler than listing the statements as text. Did you have something in mind?

File: sql/driver/result.go (hunk 10 above)
Comment (petermattis) [inline sql/driver/result.go]: https://golang.org/pkg/database/sql/driver/#Result :(

File: sql/driver/rows.go (hunk 2 above)
Comment (petermattis) [inline sql/driver/rows.go]: https://golang.org/pkg/database/sql/driver/#Rows exposes an iterator interface. `pos` is the index of the iteration into `rows`. I'll add a comment to make this clearer.

File: sql/parser/sql.go (hunk 9 above)
Comment (petermattis) [inline sql/parser/sql.go]: Is there a comment that causes github to omit diff? My understanding is that github omit diffs based on filename patterns (e.g. `.pb.go`).

Comment (tschottdorf) [inline sql/parser/sql.go]: I think Github uses some library to determine which diffs to render, and that (among other things) looks at comments usually found at the top of auto-generated files. They won't have one for us, but maybe one specific enough that we can abuse unless we want to PR it upstream.
@bdarnell knows for sure.

Comment (petermattis) [inline sql/parser/sql.go]: Looks like it isn't possible currently: https://github.com/github/linguist/issues/2206

I'm still looking.

Comment (tschottdorf) [inline sql/parser/sql.go]: @tamird to the rescue, I think: https://github.com/github/linguist/commit/d17222ec916ca8682fa77cae6a4351de004fa894

This seems to indicate that if the first line of a `.go` file includes `Code generated by`, it's considered auto-generated. Wanna give it a try?

Comment (petermattis) [inline sql/parser/sql.go]: Yep. That worked. I need to make the `go generate` command add that automatically. Should be easy.

File: sql/driver/driver_test.go (hunk 11 above)
Comment (tamird) [inline sql/driver/driver_test.go]: yeah, `testutils` is probably a better place

File: sql/parser/Makefile [hunk 12]
```diff
@@ -0,0 +1,3 @@
+sql.go: sql.y
+	go tool yacc -o sql.go sql.y
+	@echo "$$(echo '// Code generated by go yacc.' | cat - sql.go)" > sql.go
```
Comment (petermattis) [inline sql/parser/Makefile]: Let me know if there is a better/cleaner way to add a line to the top of an existing file.

File: sql/driver/driver_test.go (hunk 11 above)
Comment (petermattis) [inline sql/driver/driver_test.go]: I'll move this in a subsequent PR.

File: sql/parser/Makefile (hunk 12 above)
Comment (petermattis) [inline sql/parser/Makefile]: Hmm, that doesn't actually work. Let me try again.

Comment (petermattis) [inline sql/parser/Makefile]: Second attempt works.

Comment (bdarnell) [inline sql/parser/Makefile]: The outer `echo` looks unnecessary; is it just building up a gigantic command line? 

`awk 'BEGIN { print "// Code generated by go yacc." } { print }' < sql.go > sql.go`? `sed` can do it too but it's uglier.

File: sql/driver/conn.go [hunk 13]
```diff
@@ -0,0 +1,275 @@
...
+
+func (c *conn) Query(query string, args []driver.Value) (driver.Rows, error) {
+	stmt, err := parser.Parse(query)
+	if err != nil {
+		return nil, err
+	}
+	return c.query(stmt, args)
+}
+
+func (c *conn) Begin() (driver.Tx, error) {
```
Comment (spencerkimball) [inline sql/driver/conn.go]: Any reason you're using Tx when Txn is used elsewhere? Unless this is specifically something in SQL drivers everywhere, I'd prefer consistency.

Comment (petermattis) [inline sql/driver/conn.go]: https://golang.org/pkg/database/sql/driver/#Tx

Comment (petermattis) [inline sql/driver/conn.go]: Specifically, this method is returning a struct which implements that interface.

File: sql/parser/Makefile (hunk 12 above)
Comment (petermattis) [inline sql/parser/Makefile]: I don't think that `awk` command can work. Awk will begin outputting to the file and truncate the input. My outer echo was attempting to buffer the input, but it wasn't working properly. Switched to using a temporary file.

File: sql/driver/conn.go (hunk 1 above)
Comment (spencerkimball) [inline sql/driver/conn.go]: How about insertRow, updateRow, deleteRow and selectRows?

File: sql/driver/conn.go (hunk 3 above)
Comment (spencerkimball) [inline sql/driver/conn.go]: Document this assumption in doc.go

File: sql/driver/conn.go (hunk 1 above)
Comment (petermattis) [inline sql/driver/conn.go]: `{insert,update,delete}` can all operate on multiple rows. Should those be pluralized as well? I could also just capitalize the method names.

File: sql/driver/conn.go (hunk 3 above)
Comment (petermattis) [inline sql/driver/conn.go]: It's not an assumption: https://golang.org/pkg/database/sql/driver/#Conn

I'll add a comment.

File: sql/parser/Makefile (hunk 12 above)
Comment (bdarnell) [inline sql/parser/Makefile]: Yeah, the buffering is necessary; i was optimistically hoping that it would work if your `echo` one did. `sed -i` would be convenient here but all the ways of dealing with newlines in sed are awful and I'm not sure about getting the escaping right through a `Makefile`.

Can `go tool yacc` write to stdout?

File: sql/driver/driver.go [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+	"database/sql/driver"
+
+	"github.com/cockroachdb/cockroach/client"
+)
+
+func init() {
+	sql.Register("cockroach", &roachDriver{})
+}
+
+// roachDriver implements the database/sql/driver.Driver interface. Named
```
Comment (tamird) [inline sql/driver/driver.go]: `var _ driver = roachDriver{}`?

**Comment by spencerkimball** [comment]
LGTM

File: sql/driver/result.go (hunk 10 above)
Comment (tamird) [inline sql/driver/result.go]: womp

File: sql/driver/driver.go (hunk 14 above)
Comment (petermattis) [inline sql/driver/driver.go]: The equivalent is done at line 28.

File: Makefile [hunk 15]
```diff
@@ -119,7 +119,7 @@ acceptance:
 check:
 	errcheck -ignore 'bytes:Write.*,io:(Close|Write),net:Close,net/http:(Close|Write),net/rpc:Close,os:Close,github.com/spf13/cobra:Usage' $(PKG)
 	! go-nyet $(PKG) | grep -vE '(Weird type of StarExpr|Unknown types|`matchIndex`|`c`|cannot process directory \.git)' # TODO(tamird): https://github.com/barakmich/go-nyet/pull/10
-	! golint $(PKG) | grep -vE '(\.pb\.go|embedded\.go|yyEofCode|_string\.go)'
+	! golint $(PKG) | grep -vE '(\.pb\.go|embedded\.go|yyEofCode|_string\.go|LastInsertId)'
```
Comment (tamird) [inline Makefile]: can you add a comment above this line:
> https://golang.org/pkg/database/sql/driver/#Result :(

**Comment by tamird** [comment]
LGTM

File: sql/parser/Makefile (hunk 12 above)
Comment (petermattis) [inline sql/parser/Makefile]: Not easily. `go tool yacc -o -` writes to a file named `"-"`. I can do `go tool yacc -o /dev/stdout`, but that feels clunky. I'd prefer to keep the current code. This isn't called often, and what is there works.

File: Makefile (hunk 15 above)
Comment (petermattis) [inline Makefile]: Sure. Done.
